File: sdks/python/src/octogen_ai_sdk/operations.py
"""The client's routing table: every ``/v1`` operation this SDK can reach.

Requests name an ``operationId``, never a path. The verb and the path template
live here, once, and ``tests/contract`` checks this table against the published
OpenAPI document in both directions: a published operation with no entry is an
operation the SDK cannot reach, and an entry the contract does not define is a
request that would 404. ``POST /products/recrawl`` — shipped in both SDKs, never
a real route — is exactly what that check catches.

Templates are contract paths, not URLs: ``OctogenClient`` substitutes
``{name}`` placeholders so the value here stays comparable to the contract
verbatim.
"""

from __future__ import annotations

import re
from typing import Final, NamedTuple
from urllib.parse import quote


class Operation(NamedTuple):
    """One published operation: an HTTP verb and a contract path template."""

    method: str
    path: str


OPERATIONS: Final[dict[str, Operation]] = {
    "addUrlListUrls": Operation("POST", "/coverage/url-lists/{urlListId}/urls"),
    "checkUrlListUrls": Operation(
        "POST", "/coverage/url-lists/{urlListId}/urls/contains"
    ),
    "createUrlList": Operation("POST", "/coverage/url-lists"),
    "deleteUrlList": Operation("DELETE", "/coverage/url-lists/{urlListId}"),
    "getMe": Operation("GET", "/me"),
    "getUrlList": Operation("GET", "/coverage/url-lists/{urlListId}"),
    "getVoyage": Operation("GET", "/voyage/{task_id}"),
    "listDomains": Operation("GET", "/domains"),
    "listUrlListUrls": Operation("GET", "/coverage/url-lists/{urlListId}/urls"),
    "listUrlLists": Operation("GET", "/coverage/url-lists"),
    "listVoyages": Operation("GET", "/voyage"),
    "lookupProduct": Operation("POST", "/products/lookup"),
    "moreLikeThisProducts": Operation("POST", "/products/more-like-this"),
    "refreshProducts": Operation("POST", "/products/refresh"),
    "removeUrlListUrls": Operation(
        "POST", "/coverage/url-lists/{urlListId}/urls/remove"
    ),
    "resolveProductFromHtml": Operation("POST", "/products/resolve-from-html"),
    "searchProducts": Operation("POST", "/products/search"),
    "startVoyage": Operation("POST", "/voyage"),
}

_PLACEHOLDER = re.compile(r"\{(\w+)\}")


def resolve_operation_path(
    template: str,
    path_params: dict[str, str] | None = None,
) -> str:
    """Substitute and percent-encode ``{name}`` placeholders in ``template``."""
    params = path_params or {}

    def substitute(match: re.Match[str]) -> str:
        name = match.group(1)
        value = params.get(name)
        if not isinstance(value, str) or not value.strip():
            raise ValueError(f"{name} must be a non-empty string")
        return quote(value.strip(), safe="")

    return _PLACEHOLDER.sub(substitute, template)

File: sdks/typescript/src/operations.ts
import type { paths } from "./generated/types.js";

/**
 * The client's routing table: every `/v1` request this SDK can make.
 *
 * `path` is typed as `keyof paths` — a key of the *generated* contract — so a
 * method that names a route the published API does not define fails to compile.
 * That is the type-level half of the drift guard; the other half is
 * `tests/contract`, which fails when a published `operationId` has no entry
 * here (an operation the SDK cannot reach) or an entry names the wrong verb.
 *
 * Paths are contract templates, not URLs. `OctogenClient#request` substitutes
 * `{name}` placeholders from its `pathParams`, so the template stays
 * comparable to the contract verbatim.
 */
export type OperationPath = keyof paths;

export type HttpMethod = "DELETE" | "GET" | "PATCH" | "POST" | "PUT";

export interface OperationDefinition {
  readonly method: HttpMethod;
  readonly path: OperationPath;
}

export const OPERATIONS = {
  addUrlListUrls: { method: "POST", path: "/coverage/url-lists/{urlListId}/urls" },
  checkUrlListUrls: {
    method: "POST",
    path: "/coverage/url-lists/{urlListId}/urls/contains",
  },
  createUrlList: { method: "POST", path: "/coverage/url-lists" },
  deleteUrlList: { method: "DELETE", path: "/coverage/url-lists/{urlListId}" },
  getMe: { method: "GET", path: "/me" },
  getUrlList: { method: "GET", path: "/coverage/url-lists/{urlListId}" },
  getVoyage: { method: "GET", path: "/voyage/{task_id}" },
  listDomains: { method: "GET", path: "/domains" },
  listUrlListUrls: { method: "GET", path: "/coverage/url-lists/{urlListId}/urls" },
  listUrlLists: { method: "GET", path: "/coverage/url-lists" },
  listVoyages: { method: "GET", path: "/voyage" },
  lookupProduct: { method: "POST", path: "/products/lookup" },
  moreLikeThisProducts: { method: "POST", path: "/products/more-like-this" },
  refreshProducts: { method: "POST", path: "/products/refresh" },
  removeUrlListUrls: {
    method: "POST",
    path: "/coverage/url-lists/{urlListId}/urls/remove",
  },
  resolveProductFromHtml: { method: "POST", path: "/products/resolve-from-html" },
  searchProducts: { method: "POST", path: "/products/search" },
  startVoyage: { method: "POST", path: "/voyage" },
} as const satisfies Record<string, OperationDefinition>;

export type OperationId = keyof typeof OPERATIONS;

/** Substitute `{name}` placeholders in a contract path template. */
export function resolveOperationPath(
  template: OperationPath,
  pathParams: Readonly<Record<string, string>> = {},
): string {
  return template.replace(/\{(\w+)\}/g, (_match, name: string) => {
    const value = pathParams[name];
    if (value === undefined || value.trim().length === 0) {
      throw new TypeError(`${name} is required`);
    }
    return encodeURIComponent(value.trim());
  });
}
